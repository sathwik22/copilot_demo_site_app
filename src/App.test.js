import { render, screen, fireEvent } from '@testing-library/react';
import App from './App';

// Mock the Mermaid component since it's an external dependency
jest.mock('mermaid-react', () => ({
    __esModule: true,
    default: ({ chart }) => <div data-testid="mermaid-chart">{chart}</div>,
}));

describe('App Component', () => {
    beforeEach(() => {
        render(<App />);
    });

    // Header tests
    describe('Header', () => {
        test('renders the main heading', () => {
            expect(
                screen.getByText('GitHub Copilot Features')
            ).toBeInTheDocument();
        });
    });

    // Navigation tests
    describe('Navigation', () => {
        test('renders all navigation buttons', () => {
            expect(screen.getByText('Copilot Modes')).toBeInTheDocument();
            expect(screen.getByText('Available Models')).toBeInTheDocument();
            expect(screen.getByText('Subscription Plans')).toBeInTheDocument();
            expect(screen.getByText('Video Tutorials')).toBeInTheDocument();
        });

        test('switches content when navigation buttons are clicked', () => {
            // Initially shows Modes content
            expect(screen.getByText('Agent Mode')).toBeInTheDocument();

            // Click Models tab
            fireEvent.click(screen.getByText('Available Models'));
            expect(screen.getByText('GPT-4 Turbo')).toBeInTheDocument();

            // Click Subscriptions tab
            fireEvent.click(screen.getByText('Subscription Plans'));
            expect(screen.getByText('Copilot Enterprise')).toBeInTheDocument();

            // Click Tutorials tab
            fireEvent.click(screen.getByText('Video Tutorials'));
            expect(
                screen.getByText('Building this Website with Copilot')
            ).toBeInTheDocument();
        });
    });

    // Modes section tests
    describe('Modes Section', () => {
        beforeEach(() => {
            fireEvent.click(screen.getByText('Copilot Modes'));
        });

        test('renders all mode cards', () => {
            expect(screen.getByText('Agent Mode')).toBeInTheDocument();
            expect(screen.getByText('Edit Mode')).toBeInTheDocument();
            expect(screen.getByText('Chat Mode')).toBeInTheDocument();
        });

        test('displays mode benefits', () => {
            expect(
                screen.getByText('Automatic error handling')
            ).toBeInTheDocument();
            expect(
                screen.getByText('Multi-turn conversations')
            ).toBeInTheDocument();
            expect(
                screen.getByText('Best practices guidance')
            ).toBeInTheDocument();
        });
    });

    // Models section tests
    describe('Models Section', () => {
        beforeEach(() => {
            fireEvent.click(screen.getByText('Available Models'));
        });

        test('renders all model cards', () => {
            expect(screen.getByText('GPT-4 Turbo')).toBeInTheDocument();
            expect(screen.getByText('GPT-4')).toBeInTheDocument();
            expect(screen.getByText('GPT-3.5 Turbo')).toBeInTheDocument();
        });

        test('displays subscription badges', () => {
            expect(screen.getByText('Enterprise')).toBeInTheDocument();
            expect(screen.getByText('Pro')).toBeInTheDocument();
            expect(screen.getByText('Free')).toBeInTheDocument();
        });

        test('displays model use cases', () => {
            expect(
                screen.getByText('Complex architectural decisions')
            ).toBeInTheDocument();
            expect(
                screen.getByText('Advanced problem solving')
            ).toBeInTheDocument();
            expect(screen.getByText('Basic documentation')).toBeInTheDocument();
        });
    });

    // Subscription section tests
    describe('Subscription Section', () => {
        beforeEach(() => {
            fireEvent.click(screen.getByText('Subscription Plans'));
        });

        test('renders all subscription tiers', () => {
            expect(screen.getByText('Copilot Enterprise')).toBeInTheDocument();
            expect(screen.getByText('Copilot Pro')).toBeInTheDocument();
            expect(screen.getByText('Copilot Free')).toBeInTheDocument();
        });

        test('displays subscription features', () => {
            expect(
                screen.getByText('Access to GPT-4 Turbo')
            ).toBeInTheDocument();
            expect(screen.getByText('Higher rate limits')).toBeInTheDocument();
            expect(
                screen.getByText('Basic code completion')
            ).toBeInTheDocument();
        });
    });

    // Tutorials section tests
    describe('Tutorials Section', () => {
        beforeEach(() => {
            fireEvent.click(screen.getByText('Video Tutorials'));
        });

        test('renders all tutorial cards', () => {
            const tutorials = [
                'Building this Website with Copilot',
                'Writing Unit Tests with Copilot',
                'Code Quality & Bug Fixing',
                'Project Documentation & Architecture',
            ];

            tutorials.forEach((title) => {
                expect(screen.getByText(title)).toBeInTheDocument();
            });
        });

        test('displays tutorial metadata', () => {
            expect(screen.getByText('15:30')).toBeInTheDocument();
            expect(screen.getByText('React')).toBeInTheDocument();
            expect(screen.getByText('Tailwind CSS')).toBeInTheDocument();
        });

        test('displays tutorial chapters', () => {
            expect(
                screen.getByText('Project Setup with Copilot')
            ).toBeInTheDocument();
            expect(
                screen.getByText('UI Component Creation')
            ).toBeInTheDocument();
        });

        test('renders Mermaid diagrams for documentation tutorial', () => {
            const mermaidCharts = screen.getAllByTestId('mermaid-chart');
            expect(mermaidCharts).toHaveLength(2);
        });

        test('displays play buttons for videos', () => {
            const playButtons = screen
                .getAllByRole('button')
                .filter((button) => button.querySelector('svg'));
            expect(playButtons.length).toBeGreaterThan(0);
        });
    });
});

// Snapshot test for the entire app
describe('App Snapshots', () => {
    test('matches snapshot', () => {
        const { container } = render(<App />);
        expect(container).toMatchSnapshot();
    });
});
