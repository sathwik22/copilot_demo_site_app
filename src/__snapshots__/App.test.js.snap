// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Snapshots matches snapshot 1`] = `
<div>
  <div
    class="min-h-screen bg-gradient-to-b from-gray-50 to-gray-100"
  >
    <header
      class="bg-white shadow-sm"
    >
      <div
        class="max-w-7xl mx-auto px-4 py-6 sm:px-6 lg:px-8"
      >
        <h1
          class="text-3xl font-bold text-gray-900"
        >
          GitHub Copilot Features
        </h1>
      </div>
    </header>
    <main
      class="max-w-7xl mx-auto px-4 py-8 sm:px-6 lg:px-8"
    >
      <nav
        class="flex flex-wrap gap-4 mb-8"
      >
        <button
          class="px-4 py-2 rounded-lg font-medium bg-indigo-100 text-indigo-700"
        >
          Copilot Modes
        </button>
        <button
          class="px-4 py-2 rounded-lg font-medium text-gray-500 hover:text-gray-700"
        >
          Available Models
        </button>
        <button
          class="px-4 py-2 rounded-lg font-medium text-gray-500 hover:text-gray-700"
        >
          Subscription Plans
        </button>
        <button
          class="px-4 py-2 rounded-lg font-medium text-gray-500 hover:text-gray-700"
        >
          Video Tutorials
        </button>
      </nav>
      <div
        class="grid grid-cols-1 gap-6 sm:grid-cols-2 lg:grid-cols-3"
      >
        <div
          class="bg-white overflow-hidden shadow rounded-lg"
        >
          <div
            class="p-6"
          >
            <div
              class="flex items-center"
            >
              <div
                class="w-12 h-12 bg-indigo-100 rounded-lg flex items-center justify-center"
              >
                <svg
                  aria-hidden="true"
                  class="w-6 h-6 text-indigo-600"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m6.75 7.5 3 2.25-3 2.25m4.5 0h3m-9 8.25h13.5A2.25 2.25 0 0 0 21 18V6a2.25 2.25 0 0 0-2.25-2.25H5.25A2.25 2.25 0 0 0 3 6v12a2.25 2.25 0 0 0 2.25 2.25Z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </div>
              <h3
                class="ml-4 text-lg font-medium text-gray-900"
              >
                Agent Mode
              </h3>
            </div>
            <p
              class="mt-4 text-gray-600"
            >
              AI-powered autonomous coding sessions that handle errors and iterate until completion
            </p>
            <ul
              class="mt-4 space-y-2"
            >
              <li
                class="flex items-center text-sm text-gray-500"
              >
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 mr-2 text-indigo-500"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 18v-5.25m0 0a6.01 6.01 0 0 0 1.5-.189m-1.5.189a6.01 6.01 0 0 1-1.5-.189m3.75 7.478a12.06 12.06 0 0 1-4.5 0m3.75 2.383a14.406 14.406 0 0 1-3 0M14.25 18v-.192c0-.983.658-1.823 1.508-2.316a7.5 7.5 0 1 0-7.517 0c.85.493 1.509 1.333 1.509 2.316V18"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Automatic error handling
              </li>
              <li
                class="flex items-center text-sm text-gray-500"
              >
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 mr-2 text-indigo-500"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 18v-5.25m0 0a6.01 6.01 0 0 0 1.5-.189m-1.5.189a6.01 6.01 0 0 1-1.5-.189m3.75 7.478a12.06 12.06 0 0 1-4.5 0m3.75 2.383a14.406 14.406 0 0 1-3 0M14.25 18v-.192c0-.983.658-1.823 1.508-2.316a7.5 7.5 0 1 0-7.517 0c.85.493 1.509 1.333 1.509 2.316V18"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Multi-step task execution
              </li>
              <li
                class="flex items-center text-sm text-gray-500"
              >
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 mr-2 text-indigo-500"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 18v-5.25m0 0a6.01 6.01 0 0 0 1.5-.189m-1.5.189a6.01 6.01 0 0 1-1.5-.189m3.75 7.478a12.06 12.06 0 0 1-4.5 0m3.75 2.383a14.406 14.406 0 0 1-3 0M14.25 18v-.192c0-.983.658-1.823 1.508-2.316a7.5 7.5 0 1 0-7.517 0c.85.493 1.509 1.333 1.509 2.316V18"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Terminal output monitoring
              </li>
            </ul>
          </div>
        </div>
        <div
          class="bg-white overflow-hidden shadow rounded-lg"
        >
          <div
            class="p-6"
          >
            <div
              class="flex items-center"
            >
              <div
                class="w-12 h-12 bg-indigo-100 rounded-lg flex items-center justify-center"
              >
                <svg
                  aria-hidden="true"
                  class="w-6 h-6 text-indigo-600"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M17.25 6.75 22.5 12l-5.25 5.25m-10.5 0L1.5 12l5.25-5.25m7.5-3-4.5 16.5"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </div>
              <h3
                class="ml-4 text-lg font-medium text-gray-900"
              >
                Edit Mode
              </h3>
            </div>
            <p
              class="mt-4 text-gray-600"
            >
              Conversational, step-by-step coding experience with direct code modifications
            </p>
            <ul
              class="mt-4 space-y-2"
            >
              <li
                class="flex items-center text-sm text-gray-500"
              >
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 mr-2 text-indigo-500"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 18v-5.25m0 0a6.01 6.01 0 0 0 1.5-.189m-1.5.189a6.01 6.01 0 0 1-1.5-.189m3.75 7.478a12.06 12.06 0 0 1-4.5 0m3.75 2.383a14.406 14.406 0 0 1-3 0M14.25 18v-.192c0-.983.658-1.823 1.508-2.316a7.5 7.5 0 1 0-7.517 0c.85.493 1.509 1.333 1.509 2.316V18"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Review changes in context
              </li>
              <li
                class="flex items-center text-sm text-gray-500"
              >
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 mr-2 text-indigo-500"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 18v-5.25m0 0a6.01 6.01 0 0 0 1.5-.189m-1.5.189a6.01 6.01 0 0 1-1.5-.189m3.75 7.478a12.06 12.06 0 0 1-4.5 0m3.75 2.383a14.406 14.406 0 0 1-3 0M14.25 18v-.192c0-.983.658-1.823 1.508-2.316a7.5 7.5 0 1 0-7.517 0c.85.493 1.509 1.333 1.509 2.316V18"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Multi-turn conversations
              </li>
              <li
                class="flex items-center text-sm text-gray-500"
              >
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 mr-2 text-indigo-500"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 18v-5.25m0 0a6.01 6.01 0 0 0 1.5-.189m-1.5.189a6.01 6.01 0 0 1-1.5-.189m3.75 7.478a12.06 12.06 0 0 1-4.5 0m3.75 2.383a14.406 14.406 0 0 1-3 0M14.25 18v-.192c0-.983.658-1.823 1.508-2.316a7.5 7.5 0 1 0-7.517 0c.85.493 1.509 1.333 1.509 2.316V18"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Direct code edits
              </li>
            </ul>
          </div>
        </div>
        <div
          class="bg-white overflow-hidden shadow rounded-lg"
        >
          <div
            class="p-6"
          >
            <div
              class="flex items-center"
            >
              <div
                class="w-12 h-12 bg-indigo-100 rounded-lg flex items-center justify-center"
              >
                <svg
                  aria-hidden="true"
                  class="w-6 h-6 text-indigo-600"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7.5 8.25h9m-9 3H12m-9.75 1.51c0 1.6 1.123 2.994 2.707 3.227 1.129.166 2.27.293 3.423.379.35.026.67.21.865.501L12 21l2.755-4.133a1.14 1.14 0 0 1 .865-.501 48.172 48.172 0 0 0 3.423-.379c1.584-.233 2.707-1.626 2.707-3.228V6.741c0-1.602-1.123-2.995-2.707-3.228A48.394 48.394 0 0 0 12 3c-2.392 0-4.744.175-7.043.513C3.373 3.746 2.25 5.14 2.25 6.741v6.018Z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </div>
              <h3
                class="ml-4 text-lg font-medium text-gray-900"
              >
                Chat Mode
              </h3>
            </div>
            <p
              class="mt-4 text-gray-600"
            >
              Interactive AI assistance with context-aware responses
            </p>
            <ul
              class="mt-4 space-y-2"
            >
              <li
                class="flex items-center text-sm text-gray-500"
              >
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 mr-2 text-indigo-500"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 18v-5.25m0 0a6.01 6.01 0 0 0 1.5-.189m-1.5.189a6.01 6.01 0 0 1-1.5-.189m3.75 7.478a12.06 12.06 0 0 1-4.5 0m3.75 2.383a14.406 14.406 0 0 1-3 0M14.25 18v-.192c0-.983.658-1.823 1.508-2.316a7.5 7.5 0 1 0-7.517 0c.85.493 1.509 1.333 1.509 2.316V18"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Workspace-aware responses
              </li>
              <li
                class="flex items-center text-sm text-gray-500"
              >
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 mr-2 text-indigo-500"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 18v-5.25m0 0a6.01 6.01 0 0 0 1.5-.189m-1.5.189a6.01 6.01 0 0 1-1.5-.189m3.75 7.478a12.06 12.06 0 0 1-4.5 0m3.75 2.383a14.406 14.406 0 0 1-3 0M14.25 18v-.192c0-.983.658-1.823 1.508-2.316a7.5 7.5 0 1 0-7.517 0c.85.493 1.509 1.333 1.509 2.316V18"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Code explanation
              </li>
              <li
                class="flex items-center text-sm text-gray-500"
              >
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 mr-2 text-indigo-500"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 18v-5.25m0 0a6.01 6.01 0 0 0 1.5-.189m-1.5.189a6.01 6.01 0 0 1-1.5-.189m3.75 7.478a12.06 12.06 0 0 1-4.5 0m3.75 2.383a14.406 14.406 0 0 1-3 0M14.25 18v-.192c0-.983.658-1.823 1.508-2.316a7.5 7.5 0 1 0-7.517 0c.85.493 1.509 1.333 1.509 2.316V18"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Best practices guidance
              </li>
            </ul>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;
